base/iterators.jl

```diff
@@ -743,13 +743,27 @@ iteratoreltype(::Type{Flatten{I}}) where {I} = _flatteneltype(I, iteratoreltype(
 _flatteneltype(I, ::HasEltype) = iteratoreltype(eltype(I))
 _flatteneltype(I, et) = EltypeUnknown()
 
-flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{T}) where {T<:Tuple} = HasLength()
-flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{T}) where {T<:Number} = HasLength()
+flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{<:Tuple}) = HasLength()
+flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{<:Number}) = HasLength()
 flatten_iteratorsize(a, b) = SizeUnknown()
-flatten_length(f, ::Type{T}) where {T<:Tuple} = isleaftype(T) ? nfields(T)*length(f.it) : throw(ArgumentError("Cannot compute length of a tuple-type which is not a leaf-type"))
+
+function flatten_length(f, ::Type{<:Tuple})
+    if !isleaftype(T)
```

Comment:
ok, I thought `T::Type` may be less efficient than `::Type{T} where T`, but that it was not the case for `T::Type{<:Tuple}` which should be specialized for each different type...